### DISCUSSION TITLE
How can I get the peak memory usage for a single query?

### DISCUSSION THREAD

**Discussion by SChakravorti21** [duckdb/duckdb#10414] [Q&A]
Hi! Thanks to the DuckDB developers for making such an awesome tool.

I'm assessing the feasibility of using DuckDB for a project at work, and one of the things we need to understand is the _peak_ memory usage for individual queries. For some context:

- We're using DuckDB's C++ API to query Arrow tables.
- There might be multiple DuckDB instances within a process (e.g. one instance per thread). But if it helps, I can ensure that each instance only receives one query at a time.
- We have ~500k queries that we need to backtest and look at memory usage for, so I'm really hoping for a programmatic solution.
- Recreating the DB instance from scratch for each query _might_ be prohibitively expensive (but correct me if I'm wrong) as the queries themselves typically complete in the 10s of milliseconds.
- We want to look at memory usage specific to each query (i.e. ignoring steady-state memory usage when the system is idle).

I noticed that the `DBConfig` supports setting an `Allocator`:

https://github.com/duckdb/duckdb/blob/01d1dedf587a1e620a28c801fa18b47090e40192/src/include/duckdb/main/config.hpp#L213-L214

I was thinking I could write a proxy allocator that just keeps tracks of memory usage statistics and forwards the actual allocation requests to a real `Allocator`. But it looks like the methods on `Allocator` are generally not marked `virtual`, so maybe I will need to mark them as `virtual` when compiling DuckDB locally. This might be a problem if the codebase uses `Allocator`s by value and not by pointer/reference.

https://github.com/duckdb/duckdb/blob/01d1dedf587a1e620a28c801fa18b47090e40192/src/include/duckdb/common/allocator.hpp#L91-L93

Overall, does this approach seem reasonable? If not, is there a better solution? Any guidance is much appreciated.

**Comment by SChakravorti21** [comment]
Update: I didn't realize that the `Allocator` constructor supports taking pointers to functions that do the actual allocating/deallocating:

https://github.com/duckdb/duckdb/blob/01d1dedf587a1e620a28c801fa18b47090e40192/src/include/duckdb/common/allocator.hpp#L85-L87

It took me a while to figure out if and how I could use this to track the stats I'm interested in. In case it helps anyone else, below is roughly what I ended up implementing. If this seems like it would be useful to others, I'd be happy to make a PR 😃.

```cpp
/// Tracks statistics about memory usage.
class MemoryStats : public duckdb::PrivateAllocatorData {
  public:  /// Constructors and special member functions
    explicit MemoryStats();
    ~MemoryStats() override {}

    MemoryStats(MemoryStats&&) = delete;
    MemoryStats(const MemoryStats&) = delete;
    MemoryStats& operator=(MemoryStats&&) = delete;
    MemoryStats& operator=(const MemoryStats&) = delete;

  public:  /// Accessors and manipulators
    [[nodiscard]] constexpr auto TotalBytesAllocated() const noexcept { return total_bytes_allocated_; }
    [[nodiscard]] constexpr auto BytesInUse() const noexcept { return bytes_in_use_; }
    [[nodiscard]] constexpr auto PeakBytesInUse() const noexcept { return peak_bytes_in_use_; }

    void Increment(duckdb::idx_t size);
    void Decrement(duckdb::idx_t size);
    void Reset();

    duckdb::Allocator& GetUnderlyingAllocator() noexcept { return *underlying_; };

  private:
    std::mutex lock_{};
    duckdb::idx_t total_bytes_allocated_ = 0;
    duckdb::idx_t bytes_in_use_ = 0;
    duckdb::idx_t peak_bytes_in_use_ = 0;

    duckdb::unique_ptr<duckdb::Allocator> underlying_;
};

MemoryStats::MemoryStats()
    : duckdb::PrivateAllocatorData(),                      ///
      underlying_(duckdb::make_uniq<duckdb::Allocator>())  ///
{
}

void MemoryStats::Increment(duckdb::idx_t size)
{
    std::lock_guard guard(lock_);
    total_bytes_allocated_ += size;
    bytes_in_use_ += size;
    peak_bytes_in_use_ = std::max(bytes_in_use_, peak_bytes_in_use_);
}

void MemoryStats::Decrement(duckdb::idx_t size)
{
    std::lock_guard guard(lock_);
    bytes_in_use_ -= size;
}

void MemoryStats::Reset()
{
    std::lock_guard guard(lock_);
    total_bytes_allocated_ = 0;
    bytes_in_use_ = 0;
    peak_bytes_in_use_ = 0;
}

namespace detail {

/// Custom allocate / deallocate / reallocate functions

duckdb::data_ptr_t allocate(duckdb::PrivateAllocatorData* data, duckdb::idx_t size)
{
    auto& stats = data->Cast<MemoryStats>();
    stats.Increment(size);
    return stats.GetUnderlyingAllocator().AllocateData(size);
}

void free(duckdb::PrivateAllocatorData* data, duckdb::data_ptr_t ptr, duckdb::idx_t size)
{
    auto& stats = data->Cast<MemoryStats>();
    stats.Decrement(size);
    return stats.GetUnderlyingAllocator().FreeData(ptr, size);
}

duckdb::data_ptr_t reallocate(duckdb::PrivateAllocatorData* data,  ///
                              duckdb::data_ptr_t ptr,              ///
                              duckdb::idx_t old_size,              /// NOLINT(bugprone-easily-swappable-parameters)
                              duckdb::idx_t new_size)
{
    auto& stats = data->Cast<MemoryStats>();
    stats.Decrement(old_size);
    stats.Increment(new_size);
    return stats.GetUnderlyingAllocator().ReallocateData(ptr, old_size, new_size);
}

}  // namespace detail

duckdb::DuckDB init_duckdb()
{
    duckdb::DBConfig duckdb_config{};

    duckdb_config.allocator = duckdb::make_uniq<duckdb::Allocator>(
        detail::allocate,
        detail::free,
        detail::reallocate,
        duckdb::make_uniq<MemoryStats>()
    );

    return duckdb::DuckDB(/*filepath=*/":memory:", &duckdb_config);
}

void test_queries(std::vector<std::string> queries)
{
    duckdb::DuckDB db = init_duckdb();
    auto& stats = duckdb::Allocator::Get(*db.instance).GetPrivateData()->Cast<MemoryStats>();
    
    for (const auto& query : queries) {
        stats.Reset();
        
        // ... execute the query in DuckDB ...

        auto total_bytes_allocated = stats.TotalBytesAllocated();
        auto bytes_in_use = stats.BytesInUse();
        auto peak_bytes_in_use = stats.PeakBytesInUse();
    }
}
```

I'm only running one query at a time on a given `DuckDB` instance, and make sure to reset the stats before each query. I saw that `stats.BytesInUse()` is always exactly 0 after the query is executed, so I'm fairly certain that the bookkeeping is accurate (but correct me if I'm wrong).

**Reply by bugzpodder** [reply]
is there a way to use the class you provided in python?

**Reply by SChakravorti21** [reply]
Hey! Unfortunately, I'm not familiar enough with the DuckDB implementation to know if/how that would be possible. Not sure if it helps, but have you considered using a memory profiler like [Memray](https://github.com/bloomberg/memray)?
